src/ai/defaultai.cc

```diff
@@ -2620,22 +2640,31 @@ bool DefaultAI::construct_building(const Time& gametime) {
 	}
 	assert(persistent_data->target_military_score >= persistent_data->least_military_score);
 
-	// we must calculate wood policy
-	const Widelands::DescriptionIndex wood_index = tribe_->safe_ware_index("log");
-	// stocked wood is to be in some propotion to productionsites and
-	// constructionsites (this proportion is bit artifical, or we can say
-	// it is proportion to the size of economy). Plus some positive 'margin'
-	const int32_t stocked_wood_margin = calculate_stocklevel(wood_index) -
-	                                    productionsites.size() * 2 - numof_psites_in_constr +
-	                                    management_data.get_military_number_at(87) / 5;
-	if (gametime < Time(15 * 60 * 1000)) {
-		wood_policy_ = WoodPolicy::kAllowRangers;
-	} else if (stocked_wood_margin > 80) {
-		wood_policy_ = WoodPolicy::kDismantleRangers;
-	} else if (stocked_wood_margin > 25) {
-		wood_policy_ = WoodPolicy::kStopRangers;
-	} else {
-		wood_policy_ = WoodPolicy::kAllowRangers;
+	// we must calculate wood policy for each type of rangers
+	for (BuildingObserver& bo : rangers_) {
+		// stocked level of supported wares (only lowest counts) is to be in some proportion to
+		// productionsites and constructionsites (this proportion is bit artifical, or we can say
+		// it is proportion to the size of economy). Plus some positive 'margin'.
+		const int32_t stocked_wood_margin = get_stocklevel(bo, gametime) -
```

Comment:
this formula is exactly reused in defaultai_warfare below to replace the former woodpolicy usage